clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java

```diff
@@ -556,7 +557,7 @@ public void write(ByteBuffer buffer, Object o) {
         }
 
         @Override
-        public Records read(ByteBuffer buffer) {
+        public MemoryRecords read(ByteBuffer buffer) {
```

Comment:
We have a check above in `write` for `FileRecords`. Now that we have additional types to worry about, maybe we should change that check to to `!(o instanceof MemoryRecords)`?